core/drivers/stm32_iwdg.c

```diff
@@ -0,0 +1,400 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2022, STMicroelectronics - All Rights Reserved
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <drivers/stm32_iwdg.h>
+#include <drivers/wdt.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <kernel/pm.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <sm/sm.h>
+#include <stm32_util.h>
+#include <string.h>
+#include <trace.h>
+
+/* IWDG Compatibility */
+#define IWDG_TIMEOUT_US		U(1000)
+#define IWDG_CNT_MASK		GENMASK_32(11, 0)
+
+/* IWDG registers offsets */
+#define IWDG_KR_OFFSET		U(0x00)
+#define IWDG_PR_OFFSET		U(0x04)
+#define IWDG_RLR_OFFSET		U(0x08)
+#define IWDG_SR_OFFSET		U(0x0C)
+#define IWDG_EWCR_OFFSET	U(0x14)
+
+#define IWDG_KR_ACCESS_KEY	U(0x5555)
+#define IWDG_KR_RELOAD_KEY	U(0xAAAA)
+#define IWDG_KR_START_KEY	U(0xCCCC)
+
+/* Use a fixed prescaler divider of 256 */
+#define IWDG_PRESCALER_256	U(256)
+#define IWDG_PR_DIV_256		U(0x06)
+#define IWDG_PR_DIV_MASK	GENMASK_32(3, 0)
+
+#define IWDG_SR_PVU		BIT(0)
+#define IWDG_SR_RVU		BIT(1)
+#define IWDG_SR_WVU		BIT(2)
+#define IWDG_SR_EWU		BIT(3)
+#define IWDG_SR_UPDATE_MASK	(IWDG_SR_PVU | IWDG_SR_RVU | IWDG_SR_WVU | \
+				 IWDG_SR_EWU)
+
+#define IWDG_EWCR_EWIE		BIT(15)
+#define IWDG_EWCR_EWIC		BIT(14)
+
+/*
+ * Values for struct stm32_iwdg_device::flags
+ * IWDG_FLAGS_HW_ENABLED                Watchdog is enabled by BootROM
+ * IWDG_FLAGS_DISABLE_ON_STOP           Watchdog is freezed in SoC STOP mode
+ * IWDG_FLAGS_DISABLE_ON_STANDBY        Watchdog is freezed in SoC STANDBY mode
+ * IWDG_FLAGS_NON_SECURE                Instance is assigned to non-secure world
+ * IWDG_FLAGS_ENABLED			Watchdog has been enabled
+ */
+#define IWDG_FLAGS_HW_ENABLED			BIT(0)
+#define IWDG_FLAGS_DISABLE_ON_STOP		BIT(1)
+#define IWDG_FLAGS_DISABLE_ON_STANDBY		BIT(2)
+#define IWDG_FLAGS_NON_SECURE			BIT(3)
+#define IWDG_FLAGS_ENABLED			BIT(4)
+
+/*
+ * IWDG watch instance data
+ * @base - IWDG interface IOMEM base address
+ * @clock - Bus clock
+ * @clk_lsi - IWDG source clock
+ * @flags - Property flags for the IWDG instance
+ * @timeout - Watchdog elaspure timeout
+ * @wdt_chip - Wathcdog chip instance
+ * @link - Link in registered watchdog instance list
+ */
+struct stm32_iwdg_device {
+	struct io_pa_va base;
+	struct clk *clock;
+	struct clk *clk_lsi;
+	uint32_t flags;
+	unsigned long timeout;
```

Comment:
Why is this a long?

Reply:
this is driven by the timeout type used in _wdt.h_:
https://github.com/OP-TEE/optee_os/blob/5cbd8b3a709b1a2031f59ff10b8a60aefdea8d3b/core/include/drivers/wdt.h#L33-L35

Reply:
It's an odd choice in the interface, but it makes sense to use the same type here.